### DISCUSSION TITLE
Is there a way to access the Xterm.js `Terminal` object from an extension?

### DISCUSSION THREAD

**Discussion by ArkaneMoose** [microsoft/vscode-discussions#322] [Extension Development QnA] [answered]
I’d like to be able to access the Xterm.js [`Terminal`](https://github.com/xtermjs/xterm.js/blob/904ddffc8d370421783f0917797617bfce508854/typings/xterm.d.ts#L649-L652) object used by the integrated terminal, but I don’t see a way to access it from [the VS Code API](https://code.visualstudio.com/api/references/vscode-api). Is this possible?

The reason I’m looking for this is because I’d like to write a quick extension to listen for the [`Terminal.buffer.onBufferChange`](https://github.com/xtermjs/xterm.js/blob/904ddffc8d370421783f0917797617bfce508854/typings/xterm.d.ts#L1335-L1339) event on the active integrated terminal and invoke the [`workbench.action.toggleMaximizedPanel`](https://code.visualstudio.com/updates/v1_7#_new-commands) command accordingly.

**Comment by Drisate** [comment, marked as answer]
Yes, you can access the Xterm.js Terminal object from an extension by using the vscode.window.terminals property of the vscode module. This property is an array of Terminal objects, and you can use it to access the terminal that is currently active.

To get the active terminal, you can use the following code:

`const activeTerminal = vscode.window.terminals.find(t => t.name === vscode.window.activeTerminal.name);`

Once you have the active terminal, you can access its properties and methods, including those provided by the underlying Xterm.js Terminal object.

For example, you can use the activeTerminal.sendText method to send text to the terminal:

```
activeTerminal.sendText('echo "Hello, World!"');
You can also use the activeTerminal.onData event to listen for data that is received by the terminal:
```

```
activeTerminal.onData((data) => {
    console.log(data);
});
```

**Reply by ArkaneMoose** [reply]
Thanks for your reply, but unfortunately, this isn’t correct. The [`vscode.Terminal`](https://code.visualstudio.com/api/references/vscode-api#Terminal) objects provided through `vscode.window.terminals` or `vscode.window.activeTerminal` are not Xterm.js [`Terminal`](https://github.com/xtermjs/xterm.js/blob/904ddffc8d370421783f0917797617bfce508854/typings/xterm.d.ts#L649-L652) objects; the APIs are completely different and it seems like VS Code exposes a very limited subset of the properties and methods of an Xterm.js `Terminal`.

As I mentioned in my original post, I’m trying to access the [`Terminal.buffer.onBufferChange`](https://github.com/xtermjs/xterm.js/blob/904ddffc8d370421783f0917797617bfce508854/typings/xterm.d.ts#L1335-L1339) event, which doesn’t exist on [`vscode.Terminal`](https://code.visualstudio.com/api/references/vscode-api#Terminal).

If this isn’t possible, do I need to file a feature request on [the main VS Code repository](https://github.com/microsoft/vscode) to expose the Xterm.js [`Terminal`](https://github.com/xtermjs/xterm.js/blob/904ddffc8d370421783f0917797617bfce508854/typings/xterm.d.ts#L649-L652) object on [`vscode.Terminal`](https://code.visualstudio.com/api/references/vscode-api#Terminal)?

**Reply by worksofliam** [reply]
I can't speak for the VS Team, but the likeliness of that being exposed is very low.

**Reply by r3m0t** [reply]
@ArkaneMoose the extension is running in a separate Node process, while the xterm.js object is running in the browser, therefore the object can't be accessed. For more info - https://code.visualstudio.com/api/advanced-topics/extension-host

So, it wouldn't be possible to expose the object. You would need to file feature requests about the individual properties and events you want, eg if you want `onDidKeypress` event for the user typing.

**Reply by olguzzar** [reply]
I'm going to mark this thread as the answer. As @r3m0t mentioned, if there are specific properties or events that you want exposed, please file a feature request on the main repo with your justification for why they should be exposed. Thanks!
